[10.0] stock_available modules: optimize stock computation by avoiding to call useless compute

Cedric-Pigeon opened the pull request:
In V8.0 we used _product_available to feed a dict with data.
The problem of this method is that you compute a lot of thing several times because for each stock computed values Odoo computes the other ones.
It doesn't have big impact on small database but on biggest one it could really slow.

This PR intends to avoid these useless computations and try to benefit of the prefetches values as much as possible.


On a big database, I reduced computation time by 6.

Commit:
[CHG] optimize stock computation by avoiding to call useless compute

lasley commented on stock_available/models/product_product.py:
```diff
@@ -16,54 +16,35 @@ class ProductProduct(models.Model):
+    def _compute_available_quantities_dict(self):
+        res = {}
+        for product in self:
+            res[product.id] = {}
```
Why the individual assignments? Just create the dict with the vals:

```
res[product.id] = {
    'immediately_usable_qty': ...
```

lasley commented on stock_available/models/product_product.py:
```diff
@@ -16,54 +16,35 @@ class ProductProduct(models.Model):
+    def _compute_available_quantities(self):
+        res = self._compute_available_quantities_dict()
+        for product in self:
+            data = res[product.id]
```
Why the intermediary assignment instead of just looping `res[product.id]` directly?

lasley commented on stock_available/models/product_product.py:
```diff
@@ -16,54 +16,35 @@ class ProductProduct(models.Model):
+                if hasattr(product, key):
+                    product[key] = value
 
     immediately_usable_qty = fields.Float(
```
Please move field declarations to top - https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#models

lasley commented on stock_available/models/product_template.py:
```diff
@@ -10,75 +10,45 @@
+    def _compute_available_quantities(self):
+        res = self._compute_available_quantities_dict()
         for product in self:
+            data = res[product.id]
```
Same question regarding intermediary assignment instead of direct loop

lasley commented on stock_available/models/product_product.py:
```diff
@@ -16,54 +16,35 @@ class ProductProduct(models.Model):
-        for prod in self:
-            prod.immediately_usable_qty = res[prod.id][
-                'potential_qty']
+    def _compute_available_quantities(self):
```
It looks like this is duplicated in `product.template` why not just escalate to the template method and just pass in the `product.product` instead? Something like:

```
self.env['product.template']. _compute_available_quantities(self)
```

Admittedly I haven't tried it in this context, but I'm almost positive I've done something similar elsewhere without triggering an apples to oranges error.

lasley commented on stock_available/models/product_template.py:
```diff
@@ -10,75 +10,45 @@
-        for tmpl in self:
-            tmpl.potential_qty = res[tmpl.id]['potential_qty']
-
     immediately_usable_qty = fields.Float(
```
Move to top

lasley commented on stock_available_immediately/models/product_product.py:
```diff
@@ -3,18 +3,20 @@
         return res
+
+    @api.depends('virtual_available', 'incoming_qty')
+    def _compute_available_quantities(self):
```
What is the purpose of this method declaration? Are the depends changing or something?

Maybe add a docblock indicating the reasoning? Otherwise it's just a method calling the super.

lasley reviewed (changes requested):
Thanks @Cedric-Pigeon - comments inline

lmignon committed:
[IMP] stock_available_unreserved: Improve perf in stock computation;
Read quants without lang into the context to avoid sql join on ir.translations
Iter on product with prefetch_fields=False and lang='' to avoid reading useless column ad join on ir.translations

rousseldenis commented:
@Cedric-Pigeon Can you take into account remarks from @lasley ?

rousseldenis commented:
@Cedric-Pigeon Any news on this? Relaunched runbot and all green

rousseldenis committed:
[FIX] stock_available: Some affectations

rousseldenis committed:
[CHG] stock_available_sale: Change computation

rousseldenis commented:
@lreficent @Cedric-Pigeon @lmignon I rebased this as there is new modules that use the stock_available process.

Keep in mind that this is a change in the manner we compute available quantities. We construct a dictionary with all values and then use that dictionary to build computed ones.

That avoid to have too much confidence in how orm works for many chained computed values (but this work is based on how Odoo did the job in stock module).

MiquelRForgeFlow reviewed (approved):
LGTM :+1:

lmignon reviewed (approved):
LGTM
